decoder/src/main/java/openlr/decoder/backtracking/BackTrackingGraph.java

```diff
@@ -0,0 +1,305 @@
+     */
+    private boolean isBacktrackingRequired() {
+        return isBacktrackingRequired(currentSourceSliceIndex, graphEdge);
+    }
+
+
+    /**
+     * Forwardtracking is required if there are green successors to the current edge
+     */
+    private boolean isForwardTrackingRequired() {
+        return (graphEdge.getDestinationNode().getOutgoingEdge()
+                .stream()
+                .filter(outgoingEdge -> outgoingEdge.getEdgeColor().equals(EntityColor.GREEN)).count() != 0);
+    }
+
+
+    /**
+     * if no backtracking or forwardtracking is preformed, then next best edge is is estimated based on start or end node of the last best edge
+     */
+    private GraphEdge getNextEdgeFromNormalFlow() {
+        if (currentSourceSliceIndex < slices.size() - 1) {
+            int sourceSliceIndex = currentSourceSliceIndex;
+            int destinationSliceIndex = currentSourceSliceIndex + 1;
+            Slice sourceSlice = slices.get(sourceSliceIndex);
+            Slice destinationSlice = slices.get(destinationSliceIndex);
+
+            GraphNode destinationNode = destinationSlice.getNextBestNode();
+
+            int sourceNodeIndex = getNextSourceNodeIndex(sourceSlice);
+            int destinationNodeIndex = destinationSlice.getNodePosition(destinationNode);
+
+            String edgeid = getEdgeId(sourceSliceIndex, sourceNodeIndex, destinationSliceIndex, destinationNodeIndex);
+            graphEdge = edges.get(edgeid);
+            return graphEdge;
+        }
+        return null;
+    }
+
+    /**
+     * Backtrack the graph to identify where the connection is broken and return an unmarked edge with the end node same as the
+     * start node of the last tracked edge. if all edges are already marked then it will try to find an edge from the source slice.
+     */
+    private GraphEdge getEdgeByBackwardTracking() {
+
+        GraphEdge<T, L> edge = graphEdge;
+        GraphEdge<T, L> parent = null;
+        do {
+            parent = edge.getSourceNode().getIncomingEdge()
+                    .stream()
+                    .filter(incomingEdge -> incomingEdge.getEdgeColor().equals(EntityColor.GREEN)).findFirst().orElse(null);
+            if (parent != null) {
+                edge = parent;
+            }
+            --currentSourceSliceIndex;
+        } while (parent != null);
+
+        if (edge.getSourceNode().getIncomingEdge()
+                .stream().count() == 0) {
+            return null;
+        }
+
+        edge = edge.getSourceNode().getIncomingEdge()
+                .stream()
+                .filter(incomingEdge -> !incomingEdge.getEdgeColor().equals(EntityColor.RED)).findFirst().orElse(null);
+
+
+        if (edge == null) {
+            return getNextEdgeFromNormalFlow();
+        }
+
+        return edge;
+    }
+
+
+    /**
+     * if there are any green successor to the last edge then it will try to find the last green edge in the successor links
```

Comment:
**If** there are any green successors

Reply:
I find it really hard to understand this really long sentence.  Can you break it into smaller sentences?